Unexpected compiler behavior with TCC and array.len

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.11 d249bfb, press to see full `v doctor` output</summary>

|V full version      |V 0.4.11 d249bfb04e9c040e24213b763829ac141a811c88
|:-------------------|:-------------------
|OS                  |linux, Ubuntu 24.04.2 LTS (WSL 2)
|Processor           |6 cpus, 64bit, little endian, AMD Ryzen 5 3500 6-Core Processor
|Memory              |5.95GB/7.72GB
|                    |
|V executable        |/home/ericki/tools/v/v
|V last modified time|2025-08-04 04:59:28
|                    |
|V home dir          |OK, value: /home/ericki/tools/v
|VMODULES            |OK, value: /home/ericki/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/ericki/dev/cairo
|                    |
|Git version         |git version 2.43.0
|V git status        |d249bfb0
|.git/config present |true
|                    |
|cc version          |cc (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
|gcc version         |gcc (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
|clang version       |N/A
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (Ubuntu GLIBC 2.39-0ubuntu8.5) 2.39

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  main.v && main`
```v
module main

fn main() {
	s := 'hello 🌎'
	arr := s.runes()
	println(arr.len)
}

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/main.01K20GZ8QK0Z2A2D5MFRVKPJCG.tmp.c:1803: warning: implicit declaration of function 'array_push_noscan'
cc: /home/ericki/tools/v/thirdparty/tcc/lib/libgc.a: error: 'GC_noop1_ptr' defined twice
cc: tcc: error: undefined symbol 'array_push_noscan'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

The program should compile successfully and print `7` to the terminal.

#### **Observed behavior:**

**Scenario 1: Compiling with `tcc`**

  * The compiler fails with the error shown above.

**Scenario 2: Compiling with `gcc`**

  * The program compiles successfully and prints `7` to the terminal.

**Scenario 3: Compiling with `tcc` after adding `println(arr)`**

  * The code is modified to include `println(arr)` before `println(arr.len)`:

    ```v
    module main

    fn main() {
        s := "hello 🌎"
        arr := s.runes()
        println(arr)
        println(arr.len)
    }
    ```

  * The program now compiles successfully with `tcc`. The terminal output is:

    ```
    [`h`, `e`, `l`, `l`, `o`, ` `, `🌎`]
    7
    ```


> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/markused/walker.v
+++ b/vlib/v/markused/walker.v
@@ -634,14 +634,9 @@ fn (mut w Walker) expr(node_ ast.Expr) {
 						}
 					}
 				} else {
-					if !w.uses_append && node.op == .left_shift {
-						fsym := w.table.final_sym(node.left_type)
-						if fsym.info is ast.Array {
-							esym := w.table.final_sym(fsym.info.elem_type)
-							if esym.kind !in [.u8, .char, .string] || !w.is_direct_array_access {
-								w.uses_append = true
-							}
-						}
+					if !w.uses_append && node.op == .left_shift
+						&& w.table.final_sym(node.left_type).kind == .array {
+						w.uses_append = true
 					}
 				}
 			}

--- a/vlib/v/markused/walker.v
+++ b/vlib/v/markused/walker.v
@@ -634,8 +634,8 @@ fn (mut w Walker) expr(node_ ast.Expr) {
 						}
 					}
 				} else {
-					if !w.uses_append && node.op == .left_shift
-						&& (sym.kind == .array || w.table.final_sym(node.left_type).kind == .array) {
+					if !w.uses_append && node.op == .left_shift && (sym.kind == .array
+						|| (sym.kind == .alias && w.table.final_sym(node.left_type).kind == .array)) {
 						w.uses_append = true
 					}
 				}

--- a/vlib/v/markused/walker.v
+++ b/vlib/v/markused/walker.v
@@ -633,11 +633,10 @@ fn (mut w Walker) expr(node_ ast.Expr) {
 							w.fn_decl(mut &ast.FnDecl(opmethod.source_fn))
 						}
 					}
-				} else if sym.kind == .array {
+				} else if !w.uses_append && node.op == .left_shift && sym.kind == .array {
 					if sym.info is ast.Array {
 						esym := w.table.final_sym(sym.info.elem_type)
-						if !w.uses_append && node.op == .left_shift
-							&& (esym.kind !in [.u8, .char, .string] || !w.is_direct_array_access) {
+						if esym.kind !in [.u8, .char, .string] || !w.is_direct_array_access {
 							w.uses_append = true
 						}
 					}

--- a/vlib/v/markused/walker.v
+++ b/vlib/v/markused/walker.v
@@ -633,11 +633,14 @@ fn (mut w Walker) expr(node_ ast.Expr) {
 							w.fn_decl(mut &ast.FnDecl(opmethod.source_fn))
 						}
 					}
-				} else if !w.uses_append && node.op == .left_shift && sym.kind == .array {
-					if sym.info is ast.Array {
-						esym := w.table.final_sym(sym.info.elem_type)
-						if esym.kind !in [.u8, .char, .string] || !w.is_direct_array_access {
-							w.uses_append = true
+				} else {
+					if !w.uses_append && node.op == .left_shift {
+						fsym := w.table.final_sym(node.left_type)
+						if fsym.info is ast.Array {
+							esym := w.table.final_sym(fsym.info.elem_type)
+							if esym.kind !in [.u8, .char, .string] || !w.is_direct_array_access {
+								w.uses_append = true
+							}
 						}
 					}
 				}

--- a/vlib/v/markused/walker.v
+++ b/vlib/v/markused/walker.v
@@ -633,6 +633,14 @@ fn (mut w Walker) expr(node_ ast.Expr) {
 							w.fn_decl(mut &ast.FnDecl(opmethod.source_fn))
 						}
 					}
+				} else if sym.kind == .array {
+					if sym.info is ast.Array {
+						esym := w.table.final_sym(sym.info.elem_type)
+						if !w.uses_append && node.op == .left_shift
+							&& (esym.kind !in [.u8, .char, .string] || !w.is_direct_array_access) {
+							w.uses_append = true
+						}
+					}
 				}
 			}
 			right_type := if node.right_type == 0 && mut node.right is ast.TypeNode {
@@ -657,9 +665,6 @@ fn (mut w Walker) expr(node_ ast.Expr) {
 			}
 			if !w.uses_eq && node.op in [.eq, .ne] {
 				w.uses_eq = true
-			}
-			if !w.uses_append && node.op == .left_shift && !w.is_direct_array_access {
-				w.uses_append = true
 			}
 		}
 		ast.IfGuardExpr {

--- a/vlib/v/markused/walker.v
+++ b/vlib/v/markused/walker.v
@@ -635,7 +635,7 @@ fn (mut w Walker) expr(node_ ast.Expr) {
 					}
 				} else {
 					if !w.uses_append && node.op == .left_shift
-						&& w.table.final_sym(node.left_type).kind == .array {
+						&& (sym.kind == .array || w.table.final_sym(node.left_type).kind == .array) {
 						w.uses_append = true
 					}
 				}

--- a/vlib/v/tests/runes_test.v
+++ b/vlib/v/tests/runes_test.v
@@ -1,0 +1,7 @@
+module main
+
+fn test_main() {
+	s := 'hello'
+	arr := s.runes()
+	assert arr.len == 5
+}

